### PR TITLE
spec.py: duplicate ^dep clauses are parallel edges

### PR DESCRIPTION
**Pull request by haampie**
A concrete spec can have two nodes with identical package name thanks to `max_dupes`, so `mpileaks ^foo@1 ^foo+bar` should express two separate requirements, not one node `foo@1+bar`. From the solver's perspective, `^` can reach into two distinct unification sets: the link/run closure and direct build deps. Both package names and virtual names are unique to their unification set, meaning that for `^` neither package names nor virtuals are unique.

On develop the two clauses are merged into one node, which breaks the identities of the meet-semilattice that specs form under `satisfies`: the meet `x ∧ y` is reported empty for pairs with a common lower bound, `x ∧ x ≠ x`, `x ∧ y ≠ y ∧ x`, and the meet that is produced can be strictly below the greatest lower bound or fail to parse back. Found with the property tests of #52836.

Fix a bug where two `^` constraints on one package name are reported disjoint although a spec with two nodes satisfies both:

```python
>>> Spec("pkg-a ^foo@1.2").intersects("pkg-a ^foo@2.0")
False  # wrong
>>> Spec("pkg-a ^pkg-b %gcc@5").intersects("pkg-a ^pkg-b %gcc@6")
False  # wrong
>>> Spec("pkg-a ^foo@1").constrained("pkg-a ^foo@2")
UnsatisfiableDependencySpecError: pkg-a ^foo@2 does not satisfy pkg-a ^foo@1
```

Fix a bug where the parser merges duplicate `^dep` clauses into a single node, and rejects them when their constraints or dependency types conflict:

```python
>>> Spec("pkg-a ^foo@1 ^foo+bar")
pkg-a ^foo@1+bar  # wrong: forces one node
>>> Spec("pkg-a ^foo@1 ^foo@2")
SpecParsingError: Cannot depend on incompatible specs 'foo@1' and 'foo@2'
>>> Spec("pkg-a ^[deptypes=link] foo ^[deptypes=run] foo")
SpecParsingError: foo is a duplicate dependency, with conflicting dependency types
```

Fix a bug where merging a single edge into a build/link pair of parallel edges invents a `build,link` edge neither operand required, and the result does not parse back:

```python
>>> s = Spec("pkg-a ^[deptypes=build] pkg-b ^[deptypes=link] pkg-b")
>>> r = s.constrained("pkg-a ^[deptypes=build] pkg-b")
>>> r
pkg-a ^[deptypes=build,link] pkg-b ^[deptypes=build] pkg-b  # wrong: nothing asked for build,link
>>> Spec(str(r))
SpecParsingError: pkg-b is a duplicate dependency, with conflicting dependency types
```

Fix a bug where constrain is not idempotent, x ∧ x ≠ x:

```python
>>> s = Spec("pkg-a ^[deptypes=build] pkg-e ^[deptypes=link] pkg-e")
>>> s.constrain("pkg-a ^[deptypes=build] pkg-e ^[deptypes=link] pkg-e")
True  # wrong: nothing changed
>>> s
pkg-a ^[deptypes=build,link] pkg-e ^[deptypes=build] pkg-e
```

Fix a bug where a conditional edge is kept beside the unconditional edge that implies it:

```python
>>> Spec("%pkg-e").constrained("%[when='+bvv'] pkg-e")
%pkg-e %[when='+bvv'] pkg-e  # wrong: the unconditional edge already requires more
```

Fix a bug where two conflicting deps under one unforced condition are reported disjoint:

```python
>>> lhs, rhs = Spec("pkg-a %[when='+foo'] pkg-b@1"), Spec("pkg-a %[when='+foo'] pkg-b@2")
>>> Spec("pkg-a ~foo").satisfies(lhs) and Spec("pkg-a ~foo").satisfies(rhs)
True
>>> lhs.intersects(rhs)
False  # wrong
>>> lhs.constrained(rhs)
UnsatisfiableDependencySpecError: pkg-a %[when='+foo'] pkg-b@2 does not satisfy pkg-a %[when='+foo'] pkg-b@1
```

Fix a bug where an anonymous dependency is rejected by `constrain`, but only on the right-hand side, and two anonymous constraints are reported disjoint although a dependency per constraint satisfies both:

```python
>>> Spec("^*@2").constrained("")
^@2
>>> Spec("").constrained("^*@2")
UnconstrainableDependencySpecError: Cannot constrain by a spec containing anonymous dependencies
>>> Spec("pkg-a ^*@1").intersects("pkg-a ^*@2")
False  # wrong
```

With this change the `^` clauses stay parallel edges, and an edge merges into another only when one satisfies the other:

```python
>>> Spec("pkg-a ^foo@1 ^foo+bar")
pkg-a ^foo+bar ^foo@1
>>> Spec("pkg-a ^foo@1").constrained("pkg-a ^foo@2")
pkg-a ^foo@1 ^foo@2
>>> Spec("%pkg-e").constrained("%[when='+bvv'] pkg-e")
%pkg-e
>>> Spec("pkg-a %[when='+foo'] pkg-b@1").intersects("pkg-a %[when='+foo'] pkg-b@2")
True
```

Unconditional `%` constraints to one package name still merge, since a package has at most one direct dependency per name. Edges under `when` conditions are one node only where both conditions hold, and an edge records exactly one condition, so they stay apart. An edge implied by another edge to the same name is still merged away, so `pkg-a ^foo@1 ^foo` remains a single edge.

An anonymous constraint requires some dependency to match it, and without a name nothing forces it onto one node with another edge, so it is a parallel edge too, direct or transitive. These cases pin down that the meet is the greatest lower bound, not just any lower bound: `%+foo+bar` satisfies both `%+foo` and `%+bar`, but so does `%+foo %+bar` with a different dependency for each, so the parallel pair is the meet and the fused edge is strictly below it:

```python
>>> Spec("pkg-a %+foo").constrained("pkg-a %+bar")
pkg-a %+bar %+foo
```

Independent of edge merging, fix a bug where `constrain` assigns a name to an anonymous spec but leaves the dependent edges of its children keyed by the anonymous name:

```python
>>> s = Spec("^bar@3")
>>> bar = s.edges_to_dependencies("bar")[0].spec
>>> s.constrain(Spec("foo"))
True
>>> bar._dependents.keys()
dict_keys([''])  # wrong: the parent's name is now foo
```

When the name is assigned, those edges are now rekeyed.

Also independent of edge merging, fix a bug where `constrain` with a concrete right-hand side detaches the spec from its dependents. The concrete spec is copied over the node with `_dup`, which prunes in-edges: correct for a detached copy, wrong for a node still filed in its dependents' edge maps.

```python
>>> s = Spec("pkg-a ^pkg-b")
>>> child = s.edges_to_dependencies("pkg-b")[0].spec
>>> child.constrain(concretize_one("pkg-b"))
True
>>> child.dependents()
[]  # wrong: pkg-a still depends on it
```

The in-edges are now kept when the node's contents are replaced.

### CODE REVIEW

**Comment by haampie** [comment]
Notice that in Spack v1.0 the following

```
^[deptypes=link] foo@1 ^[deptypes=build] foo@2
```

already parses as two distinct nodes, so this PR just generalizes it.

File: lib/spack/spack/spec.py [hunk 1]
```diff
@@ -2119,8 +2091,85 @@ def add_dependency_edge(
...
+            return changed
+
+        if not owned:
+            candidate.spec = candidate.spec.copy(deps=True)
+        _add_edge_to_map(self._dependencies, candidate.spec.name, candidate)
+        _add_edge_to_map(candidate.spec._dependents, self.name, candidate)
+        return True
+
+    def _detach_edge(self, edge: DependencySpec) -> None:
+        """Remove an edge from this spec and from the dependents of the node it points at."""
```
Comment (alalazo) [inline lib/spack/spack/spec.py]: Both assignments here can assign the empty list. This can lead to subtle failures like the following:
```python
>>> from spack.spec import Spec
>>> s = Spec("pkg-a ^foo")
>>> child = s["foo"]
>>> s.constrain(Spec("pkg-a ^foo@1"))
True
>>> print(child._dependents)
{'pkg-a': []}
>>> print(child.root)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 2198, in root
    return edges_by_package[0].parent.root
           ~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

File: lib/spack/spack/spec.py [hunk 2]
```diff
@@ -3365,13 +3407,22 @@ def _intersects_dependencies(self, other, resolve_virtuals: bool = True):
...
-        # Handle first-order constraints directly
-        common_dependencies = {x.name for x in self.dependencies()}
-        common_dependencies &= {x.name for x in other.dependencies()}
-        for name in common_dependencies:
-            if not self[name]._intersects(
-                other[name], deps=True, resolve_virtuals=resolve_virtuals
-            ):
+        # A package has at most one direct dependency per name, so a concretization satisfies
+        # all unconditional direct edges to one name with a single node. Transitive constraints
+        # to one name never conflict with each other: a DAG can hold two nodes of the same name.
```
Comment (alalazo) [inline lib/spack/spack/spec.py:3430]: Should we try to reason on conditional edges / simplify "when" conditions that are satisfied by the abstract spec already? This example for instance seems wrong:
```python
>>> from spack.spec import Spec
>>> l = Spec("pkg-a+foo %[when='+foo'] pkg-b@1")
>>> r = Spec("pkg-a+foo %[when='+foo'] pkg-b@2")
>>> print(l.intersects(r))
True
>>> print(l.constrained(r))
pkg-a+foo %[when='+foo'] pkg-b@1 %[when='+foo'] pkg-b@2
```
if we assume at most one direct dependency per package name. Are there trade-offs to be made for these cases?

File: lib/spack/spack/spec.py
Comment (alalazo) [inline lib/spack/spack/spec.py:3423]: Do we need some early return cases in this function for specs of the form `abstract %foo/hash`? The following snippet shows we are copying nodes from concrete subdags:
```python
>>> import gzip
>>> import spack.deptypes as dt
>>> from spack.spec import Spec
>>> # A concrete 37-node child under an abstract root (what "^/hash" gives you after lookup_hash)
>>> with gzip.open("lib/spack/spack/test/data/specfiles/hdf5.v020.json.gz", "rt") as f:
...     child = Spec.from_json(f.read())
... 
>>> def parent(c):
...     p = Spec("pkg-a")
...     p.add_dependency_edge(c, depflag=dt.BUILD, virtuals=(), direct=True)
...     return p
... 
>>> copied, orig = [], Spec.copy
>>> def counting_copy(self, deps=True, **kw):
...     copied.append(self)
...     return orig(self, deps=deps, **kw)
... 
>>> Spec.copy = counting_copy
>>> l, r = parent(child), parent(child.copy())
>>> copied.clear()
>>> l._intersects(r, deps=True, resolve_virtuals=False)   # resolve_virtuals=False: skip ProviderIndex
True
>>> Spec.copy = orig
>>> print("copies:", len(copied))
copies: 37
>>> print("root of the copied subtree:", copied[0].name, "concrete:", copied[0].concrete)
root of the copied subtree: hdf5 concrete: True
```

File: lib/spack/spack/spec.py [hunk 3]
```diff
@@ -3365,13 +3407,22 @@ def _intersects_dependencies(self, other, resolve_virtuals: bool = True):
...
+        groups: Dict[str, List[DependencySpec]] = {}
+        for edge in itertools.chain(self.edges_to_dependencies(), other.edges_to_dependencies()):
+            if edge.direct and edge.spec.name and edge.when is EMPTY_SPEC:
+                groups.setdefault(edge.spec.name, []).append(edge)
+
+        for edges in groups.values():
+            if len(edges) < 2:
+                continue
+            merged = edges[0].spec.copy(deps=True)
```
Comment (alalazo) [inline lib/spack/spack/spec.py:3450]: Related to the comment on `_intersects_dependencies`, do we need to copy if `edge.spec.concrete is True`?

File: lib/spack/spack/test/concretization/core.py [hunk 4]
```diff
@@ -5610,3 +5610,15 @@ def test_terminal_ui_announces_pool_only_when_solving(total, announced, capsys):
     )
 
     assert ("Starting concretization" in capsys.readouterr().out) is announced
+
+
+def test_parallel_edges_in_a_literal_reach_the_solver(mock_packages, config):
+    """A duplicate ^dep clause parses as parallel edges rather than one merged node. The solver
```
Comment (alalazo) [inline lib/spack/spack/test/concretization/core.py:5678]: 👍 Ok, this tests the current state, and need to be relaxed when time comes to extend the model.

File: lib/spack/spack/test/spec_dag.py [hunk 5]
```diff
@@ -1113,3 +1121,32 @@ def test_getitem_finds_transitive_virtual():
     x.add_dependency_edge(y, depflag=dt.LINK, virtuals=())
     y.add_dependency_edge(z, depflag=dt.LINK, virtuals=("virtual",))
     assert x["virtual"].name == "z"
+
+
+def test_copy_preserves_parallel_edges_with_identical_attributes(mock_packages):
+    """Two parallel edges to one name that agree on every attribute _satisfies_edge()
```
Comment (alalazo) [inline lib/spack/spack/test/spec_dag.py]: Can we ever reach the synthetic case for this branch? Wouldn't it be better to use something like:
```
s = Spec("pkg-a ^pkg-b %pkg-c ^pkg-b %pkg-d")
``` 
that keeps the edges separate?

File: lib/spack/spack/test/spec_semantics.py [hunk 6]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    backward = conditional.constrained(unconditional)
+    assert forward == unconditional
+    assert backward == unconditional
+    assert forward.to_dict() == backward.to_dict()
+
+
+def test_edges_differing_in_namespace_stay_parallel(mock_packages):
+    """`^pkg-b@1` does not satisfy `^builtin_mock.pkg-b` and vice versa: a concretization can
+    hold a builtin_mock node next to a pkg-b@1 node from another repo, so the merge keeps both
+    edges. An edge that does carry the namespace absorbs the one it satisfies."""
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: I don't think concretization can do that right now, but I agree with the semantics here. I'd just adjust the docstring to avoid overclaiming on concretization capabilities.

File: lib/spack/spack/test/spec_semantics.py [hunk 7]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    """Duplicate % clauses to one name merge into a single direct edge that keeps the propagated
+    policy, whichever clause states it."""
+    assert Spec("mpileaks %callpath %%callpath") == Spec("mpileaks %%callpath")
+    assert Spec("mpileaks %%callpath %callpath") == Spec("mpileaks %%callpath")
+
+
+def test_satisfies_ignores_edge_propagation(mock_packages):
+    """%% expresses a preference the solver may override, so it does not narrow the set of DAGs:
+    satisfaction ignores it in both directions. In particular concrete specs, whose edges record
+    no propagation, satisfy a %% input."""
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: Here we talk about concrete specs but we don't test them.

File: lib/spack/spack/test/spec_semantics.py [hunk 8]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    assert Spec("mpileaks %%callpath").satisfies("mpileaks %callpath")
+    assert Spec("mpileaks %callpath").satisfies("mpileaks %%callpath")
+
+
+def test_conditional_propagated_edge_is_not_redundant(mock_packages):
+    """A conditionally propagated edge is satisfied by an unconditional plain one, but it states
+    a preference the plain edge does not, so the two stay parallel instead of the propagated edge
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: (nitpick) Here we don't check the satisfied part (which is checked above, but makes the test docs a bit inconsistent)?

File: lib/spack/spack/test/spec_semantics.py [hunk 9]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    spec = Spec("mpileaks %[virtuals=c] llvm ^[virtuals=c] gcc")
+    edges = spec.edges_to_dependencies()
+    assert len(edges) == 2
+    assert {edge.spec.name for edge in edges} == {"llvm", "gcc"}
+
+
+def test_two_providers_of_one_virtual_merge_as_parallel_edges(mock_packages):
+    """Two edges naming different providers of one virtual are two requirements, each matched
+    anywhere in the DAG. The parser and constrain keep them side by side."""
+    spec = Spec("mpileaks ^[virtuals=mpi] mpich ^[virtuals=mpi] zmpi")
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: I think this is not true if we go through the provider index path? Not sure we should fix it here, but worth at least noting:
```python
>>> from spack.spec import Spec
>>> import spack.provider_index, spack.repo
>>> l, r = Spec("mpileaks ^mpi@3:"), Spec("^mpich@1.0")
>>> print("same-name edges  :", Spec("mpileaks ^mpich@1.2").intersects(Spec("^mpich@2.0")))
same-name edges  : True
>>> print("virtual/provider :", l.intersects(r))
virtual/provider : False
>>> print("  without virtuals:", l._intersects(r, deps=True, resolve_virtuals=False))
  without virtuals: True
```

File: lib/spack/spack/test/spec_semantics.py [hunk 10]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    lhs, rhs = Spec("pkg-a ^[virtuals=mpi] mpich"), Spec("pkg-a ^[virtuals=mpi] zmpi")
+    assert lhs.intersects(rhs) and rhs.intersects(lhs)
+    forward, backward = lhs.constrained(rhs), rhs.constrained(lhs)
+    assert len(forward.edges_to_dependencies()) == 2
+    assert forward.to_dict() == backward.to_dict()
+
+
+def test_two_providers_under_conditions_that_exclude_each_other_are_fine(mock_packages):
+    """Only one provider can be the one at a time, so two of them named under conditions that
+    cannot hold together are not in each other's way."""
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py:2940]: Doesn't this docstring contradict the previous one? Do we need this test?

File: lib/spack/spack/test/spec_semantics.py [hunk 11]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    # the same union from the other side: the pair absorbs the lone build edge
+    backward = Spec("pkg-a ^[deptypes=build] pkg-b ^[deptypes=link] pkg-b").constrained(
+        Spec("pkg-a ^[deptypes=build] pkg-b")
+    )
+    assert backward.to_dict() == result.to_dict()
+
+
+def test_parallel_direct_edges_are_always_the_same_edge(mock_packages):
+    """A package has at most one direct dependency on a given name, so two direct edges to one
+    name are always merged into one, whatever their deptypes."""
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py:2970]: Per a comment before, it seems we don't merge conditional edges when they have the same condition? Is that wanted?

File: lib/spack/spack/test/spec_semantics.py [hunk 12]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+
+    edges = forward.edges_to_dependencies()
+    assert len(edges) == 1
+    assert edges[0].virtuals == ("c", "cxx")
+    assert edges[0].spec.satisfies("gcc@5")
+
+
+def test_two_versions_of_one_provider_of_a_virtual_intersect(mock_packages):
+    """Two edges naming the same provider for one virtual at versions that do not intersect are
+    still two existential requirements, each matched by its own node."""
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: "existential requirements" 🤔 This sounds more metaphysical than it needs to be 😆

File: lib/spack/spack/test/spec_semantics.py [hunk 13]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    assert spec.intersects(spec)
+    assert spec.constrain("pkg-a ^*@2") is False
+
+    # commutativity: the meet is the same from either side
+    forward, backward = Spec("").constrained(spec), spec.constrained("")
+    assert forward.to_dict() == backward.to_dict() == spec.to_dict()
+
+    # two anonymous constraints can each be matched by a different dependency, so they stay two
+    # edges, transitive or direct. The single edge "^+foo+bar" also satisfies both operands, but
+    # the meet is the greatest lower bound: the witness with two distinct deps satisfies both
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: witness?

File: lib/spack/spack/test/spec_semantics.py [hunk 14]
```diff
@@ -2834,3 +2837,301 @@ def test_mark_concrete_roundtrip_preserves_hashes(spec_str, config, mock_package
...
+    # a named edge alone satisfies both requirements of a named/anonymous pair
+    named, anonymous = Spec("pkg-a ^pkg-b@2"), Spec("pkg-a ^*@2")
+    forward, backward = named.constrained(anonymous), anonymous.constrained(named)
+    assert forward.to_dict() == backward.to_dict()
+    assert forward.satisfies(named) and named.satisfies(forward)
+
+
+def test_an_edge_bridging_two_parallel_edges_relates_nothing(mock_packages):
+    """One edge listing two virtuals does not weld the edges listing them separately into one
+    node: all three are existential requirements, each matched by its own node."""
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: weld / existential requirements ?

File: lib/spack/spack/test/spec_syntax.py [hunk 15]
```diff
@@ -716,6 +717,52 @@ def _specfile_for(spec_str, filename):
...
+                Token(SpecTokens.START_EDGE_PROPERTIES, value="^["),
+                Token(SpecTokens.KEY_VALUE_PAIR, value="deptypes=build"),
+                Token(SpecTokens.END_EDGE_PROPERTIES, value="]"),
+                Token(SpecTokens.UNQUALIFIED_PACKAGE_NAME, value="zlib"),
+            ],
+            "^[deptypes=link] zlib ^[deptypes=build] zlib",
+        ),
+        # Neither depflag is a superset of the other, so the edges stay parallel, as in the
+        # link/build case above. Indirect edges to one name are never one node on deptypes
+        # alone; a shared virtual, or being direct, is what fuses them.
```
Comment (alalazo) [inline lib/spack/spack/test/spec_syntax.py]: Why the comment is between two seemingly identical tests? Did I miss some difference in the structure?

**Review by alalazo — COMMENTED** [review]
There are many bugs fixed here, so 👍 I tried to find inconsistencies in the new semantics, and left a few comments below. We can probably leave most of these issues for later, but it's worth verifying which are still to be considered "bugs".

I think some weird LLMs naming went in ("weld" instead of "merge", or "existential requirements" for constraints)

File: lib/spack/spack/spec.py (hunk 2 above)
Comment (haampie) [reply lib/spack/spack/spec.py:3430]: Statically evaluating the when condition and turning conditional into unconditional edges (and then simplifying and raising if disjoint) is what happens in my other branch. It's non-trivial cause it can also happen during parsing where the node is still under construction (e.g. `Spec("%[when=%foo] bar %foo")`. I purposefully left it out here.

File: lib/spack/spack/test/concretization/core.py (hunk 4 above)
Comment (haampie) [reply lib/spack/spack/test/concretization/core.py:5678]: This particular test will pass also if the solver is relaxed because `mpich` doesn't allow dupes, but yes, if/once the solver encoding is relaxed, a test can be added to assert that `^foo@1 ^foo@2` is not unified, and even if so, the root needs a pure build dep on mpich for it to work.

File: lib/spack/spack/test/spec_semantics.py (hunk 9 above)
Comment (haampie) [reply lib/spack/spack/test/spec_semantics.py]: That's material of the other PR that makes it stateless.

File: lib/spack/spack/test/spec_semantics.py (hunk 10 above)
Comment (haampie) [reply lib/spack/spack/test/spec_semantics.py:2940]: That was supposed to be `%`

File: lib/spack/spack/test/spec_semantics.py (hunk 11 above)
Comment (haampie) [reply lib/spack/spack/test/spec_semantics.py:2970]: Yes, desirable, but this PR would be much larger when dealing with all `when` edge cases.

File: lib/spack/spack/test/spec_syntax.py (hunk 15 above)
Comment (haampie) [reply lib/spack/spack/test/spec_syntax.py]: I think it's a reference to develop's `spack.deptypes.compatible` which of course no longer exist and is practically impossible to return as a regression. I've deleted it.

File: lib/spack/spack/test/spec_semantics.py (hunk 7 above)
Comment (haampie) [reply lib/spack/spack/test/spec_semantics.py]: The comment was meant to say that if you pass `%%foo` input to the solver you get `%foo` output, so `%foo.satisfies(%%foo)` must hold to avoid "output does not satisfy input". However, I think the better solution is to branch on lhs concrete for things related to propagation, so I'll just drop the comment for now.

File: lib/spack/spack/spec.py [hunk 16]
```diff
@@ -2167,17 +2167,12 @@ def _detach_edge(self, edge: DependencySpec) -> None:
...
-        for name, dependents in edge.spec._dependents.items():
-            if any(e is edge for e in dependents):
-                remaining = [e for e in dependents if e is not edge]
-                if remaining:
-                    edge.spec._dependents[name] = remaining
-                else:
-                    del edge.spec._dependents[name]
-                break
+        dependents = edge.spec._dependents
+        remaining = [e for e in dependents[self.name] if e is not edge]
```
Comment (haampie) [inline lib/spack/spack/spec.py:2171]: @alalazo can you merge #52869 so I don't have to think about `x is y`?

**Review by alalazo — COMMENTED** [review]
I think the latest changes introduced a regression:
```python
>>> import gzip
>>> import traceback
>>> 
>>> import spack.deptypes as dt
>>> from spack.spec import Spec
>>> with gzip.open("lib/spack/spack/test/data/specfiles/hdf5.v020.json.gz", "rt") as f:
...     concrete = Spec.from_json(f.read())
... 
>>> s = Spec("%hdf5")  # anonymous root, one direct edge to an abstract hdf5
>>> t = Spec("")
>>> t.add_dependency_edge(concrete, depflag=dt.BUILD, virtuals=(), direct=True)
>>> s.constrain(t)
True
>>> child = s.edges_to_dependencies()[0].spec
>>> child.concrete
True
>>> child.root.name  # pre-existing, so not something we need to fix here
'hdf5'
>>> child._dependents
{}
>>> s.constrain(Spec("zlib"))
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 3159, in constrain
    return self._constrain(other, deps=deps, resolve_virtuals=True)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 3246, in _constrain
    remaining = [e for e in dependents[""] if e is not edge]
                            ~~~~~~~~~~^^^^
KeyError: ''
```

File: lib/spack/spack/spec.py (hunk 2 above)
Comment (alalazo) [reply lib/spack/spack/spec.py:3430]: Ok, let's deal with that in a following PR. I don't think people will hit this one often if it's not consistent right now.

File: lib/spack/spack/test/spec_semantics.py (hunk 9 above)
Comment (alalazo) [reply lib/spack/spack/test/spec_semantics.py]: Ok, same as above: I don't think people will hit this frequently in the window between the two merges, so fine with me to postpone.

**Comment by haampie** [comment]
Ok, that issue seems to be related to "abuse" of `_dup` on deps, cause it prunes self's in-edges. I guess that's justified, cause there's a 1-1 relationship between parent and child that would turn into 1-many if the duplicated child retains its in edges.

The same bug exists on develop, cause it also uses `_dup` on deps:

```python
s, t = Spec("%hdf5"), Spec("")
t.add_dependency_edge(concrete_hdf5, depflag=dt.BUILD, virtuals=(), direct=True)
s.constrain(t)
edge = s.edges_to_dependencies()[0]
edge.spec._dependents  # is {} thanks to _dup
```

**Review by alalazo — APPROVED** [review]

**Comment by haampie** [comment]
E4S concretization did not change as a result of this PR.

**Comment by haampie** [comment]
Removing the v1.0.5, v1.1.2 backport labels.

Part of the `spec.py` algebra stack; applying it here would require most of develop's `spec.py` evolution.

It is still on `develop`; only the backport is being dropped. Re-label it if someone
wants to port it to the release branch by hand.
